examples/src/components/BucketExecutorExample.jsx

```diff
@@ -0,0 +1,250 @@
+        },
+    },
+    onDemandExecution: {
+        props: {
+            dimensions: [[Model.measure(totalSalesIdentifier)]],
+            autoLoadFirstPage: false,
+        },
+        label: "On Demand Execution",
+        // eslint-disable-next-line react/prop-types
+        DataComponent: ({ result, getPage }) => {
+            return (
+                <div
+                    className="s-bucket-executor-kpi"
+                    style={{
+                        display: "flex",
+                        flexDirection: "column",
+                        alignItems: "flex-start",
+                        minHeight: "3em",
+                        lineHeight: "1em",
+                        margin: "10px 0",
+                        fontSize: 30,
+                        verticalAlign: "bottom",
+                        fontWeight: 700,
+                    }}
+                >
+                    {result ? result.data[0] : <p>No data</p>}
+                    <div>
+                        <button
+                            onClick={() => getPage({ offset: [0, 0], limit: [1, 1] })}
+                            className="button button-action s-retry-button"
+                        >
+                            Get Data
+                        </button>
+                    </div>
+                </div>
+            );
+        },
+    },
+};
+
+export class BucketExecutorExample extends Component {
+    constructor(props) {
+        super(props);
+
+        this.state = {
+            selectedPresetKey: Object.keys(presets)[0],
+        };
+    }
+
+    setPreset = selectedPresetKey => {
+        this.setState({
+            selectedPresetKey,
+        });
+    };
+
+    render() {
+        const { selectedPresetKey } = this.state;
+        const { props: presetProps, DataComponent } = presets[selectedPresetKey];
+
+        return (
+            <div>
+                {/*
+                    We need to render the Execute component even in loading
+                    otherwise the ongoing request is cancelled
+                */}
+                <p>
+                    {Object.keys(presets).map(presetKey => {
+                        const preset = presets[presetKey];
+                        return (
+                            <button
+                                key={presetKey}
+                                onClick={() => this.setPreset(presetKey)}
+                                className="button button-action s-retry-button"
+                            >
+                                {preset.label}
+                            </button>
+                        );
+                    })}
+                </p>
+                <BucketExecutor
+                    projectId={projectId}
+                    {...presetProps}
+                >
+                    {childrenProps => {
+                        const { result, response, isLoading, error } = childrenProps;
+
+                        if (error) {
```

Comment:
This is a matter of personal taste but consider having these as a props: `ErrorComponent`, `LoadingComponent` and either `DataComponent` or make that the render prop. The advantage of this approach is twofold

1. You don't force the user to make the same `if...else` in every component (arguably, every use will handle the three states by rendering distinct components).
1. You can provide defaults for the `ErrorComponent` and/or `LoadingComponent`

This is how `react-native`'s built-ins often handle this and in my experience this is more pleasant to work with, see their basic list implementation – [FlatList](https://github.com/facebook/react-native/blob/b3a50ec0de1065c9fdc6bc8453b8534a4498058a/Libraries/Lists/FlatList.js#L78)

Reply:
Updated